riscv-rt/macros/src/lib.rs

```diff
@@ -52,29 +52,57 @@ use proc_macro::TokenStream;
 pub fn entry(args: TokenStream, input: TokenStream) -> TokenStream {
     let f = parse_macro_input!(input as ItemFn);
 
+    #[cfg(not(feature = "u-boot"))]
+    let arguments_limit = 3;
+    #[cfg(feature = "u-boot")]
+    let arguments_limit = 2;
+
     // check the function arguments
-    if f.sig.inputs.len() > 3 {
+    if f.sig.inputs.len() > arguments_limit {
         return parse::Error::new(
             f.sig.inputs.last().unwrap().span(),
             "`#[entry]` function has too many arguments",
         )
         .to_compile_error()
         .into();
     }
-    for arg in &f.sig.inputs {
-        match arg {
-            FnArg::Receiver(_) => {
-                return parse::Error::new(arg.span(), "invalid argument")
-                    .to_compile_error()
-                    .into();
-            }
-            FnArg::Typed(t) => {
-                if !is_simple_type(&t.ty, "usize") {
-                    return parse::Error::new(t.ty.span(), "argument type must be usize")
-                        .to_compile_error()
-                        .into();
-                }
-            }
+
+    fn check_simple_type(argument: &PatType, ty: &str) -> Option<TokenStream> {
+        let inv_type_message = format!("argument type must be {}", ty);
+
+        if !is_simple_type(&argument.ty, ty) {
+            let error = parse::Error::new(argument.ty.span(), inv_type_message);
+
+            Some(error.to_compile_error().into())
+        } else {
+            None
+        }
+    }
+    fn check_argument_type(argument: &FnArg, ty: &str) -> Option<TokenStream> {
+        let argument_error = parse::Error::new(argument.span(), "invalid argument");
+        let argument_error = argument_error.to_compile_error().into();
+
+        match argument {
+            FnArg::Typed(argument) => check_simple_type(argument, ty),
+            FnArg::Receiver(_) => Some(argument_error),
+        }
+    }
+    #[cfg(not(feature = "u-boot"))]
+    for argument in f.sig.inputs.iter() {
+        if let Some(message) = check_argument_type(argument, "usize") {
+            return message;
+        };
+    }
+    #[cfg(feature = "u-boot")]
+    if let Some(argument) = f.sig.inputs.get(0) {
+        if let Some(message) = check_argument_type(argument, "c_int") {
+            return message;
+        }
+    }
+    #[cfg(feature = "u-boot")]
+    if let Some(argument) = f.sig.inputs.get(1) {
+        if let Some(message) = check_argument_type(argument, "usize") {
```

Comment:
Isn't the second argument a `char *const argv[]`? Shouldn't we expect something like `*const *const c_char`?

Reply:
Yes, it is. I will change it to *const *const c_char.